simplified-ui-announcements/src/main/java/org/nypl/simplified/ui/announcements/AnnouncementsController.kt

```diff
@@ -0,0 +1,176 @@
+
+  private var profile: ProfileReadableType? = null
+  private var subscriptions = CompositeDisposable()
+  private val handlingUpdate = AtomicBoolean(false)
+
+  @Volatile
+  private var lifecycleOwner: LifecycleOwner? = null
+
+  @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
+  fun onViewAvailable(owner: LifecycleOwner) {
+    this.lifecycleOwner = owner
+    this.subscriptions = CompositeDisposable()
+    this.subscriptions.add(this.profileController.accountEvents().subscribe(this::onAccountEvent))
+    this.subscriptions.add(this.profileController.profileEvents().subscribe(this::onProfileEvent))
+    this.profile = this.profileController.profileCurrent()
+  }
+
+  @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
+  fun onViewUnavailable(owner: LifecycleOwner) {
+    this.lifecycleOwner = owner
+    this.subscriptions.dispose()
+    this.profile = null
+  }
+
+  private fun onAccountEvent(event: AccountEvent) {
+    return when (event) {
+      is AccountEventUpdated -> {
+        checkForAnnouncements()
+      }
+      else -> {
+        // Nothing to do
+      }
+    }
+  }
+
+  /*
+   * Check to see if there are any unread announcements on whatever is the most recent account.
+   */
+
+  private fun checkForAnnouncements() {
+    if (this.handlingUpdate.compareAndSet(false, true)) {
+      try {
+        val profileNow = this.profile ?: return
+        val accountViewing = profileNow.preferences().mostRecentAccount ?: return
+        val accountUpdated = profileNow.account(accountViewing)
+        this.tryPublishingAnnouncements(accountUpdated)
+      } finally {
+        this.handlingUpdate.set(false)
+      }
+    }
+  }
+
+  private fun tryPublishingAnnouncements(account: AccountType) {
+    val acknowledged =
+      account.preferences.announcementsAcknowledged.toSet()
+    val notYetAcknowledged =
+      account.provider.announcements.filter { !acknowledged.contains(it.id) }
+
+    if (notYetAcknowledged.isNotEmpty()) {
+      this.showDialogForAnnouncements(account, notYetAcknowledged)
+      account.setPreferences(
+        account.preferences.copy(
+          announcementsAcknowledged = account.provider.announcements.map(Announcement::id)
+        )
+      )
+    }
+  }
+
+  private fun showDialogForAnnouncements(
+    account: AccountType,
+    announcements: List<Announcement>
+  ) {
+    this.uiThread.runOnUIThread {
+      val owner = this.lifecycleOwner ?: return@runOnUIThread
+      if (owner.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
+        showDialogForAnnouncementsUI(account, announcements)
+      }
+    }
+  }
+
+  @UiThread
+  private fun showDialogForAnnouncementsUI(
+    account: AccountType,
+    announcements: List<Announcement>
+  ) {
+    var indexNumber = 0
+
+    val title =
+      context.getString(
+        R.string.announcementTitle,
+        account.provider.displayName,
+        indexNumber + 1,
+        announcements.size
+      )
+
+    val dialog =
```

Comment:
I don't think a dialog is an ideal ux for this. Instead, I'd suggest we show a snackbar with the text `%d unread announcements`. When the user taps the snackbar we push a new fragment onto the stack that displays the announcements.

Reply:
I agree, I'd rather have a separate UI entirely that allows for reading a scrolling list of announcements. They just said dialog box, though... :slightly_frowning_face: